pkg/controller/metrics/provision_underway_collector.go

```diff
@@ -0,0 +1,91 @@
+package metrics
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1"
+	controllerutils "github.com/openshift/hive/pkg/controller/utils"
+)
+
+var (
+	// ClusterDeployment conditions that could indicate provisioning problems with a cluster
+	provisioningDelayCondition = [...]hivev1.ClusterDeploymentConditionType{
+		hivev1.ProvisionFailedCondition,
+		hivev1.InstallLaunchErrorCondition,
+	}
+)
+
+// provisioning underway metrics collected through a custom prometheus collector
+type provisioningUnderwayCollector struct {
+	client client.Client
+
+	// metricClusterDeploymentProvisionUnderwaySeconds is a prometheus metric for the number of seconds
+	// between when a still provisioning cluster was created and now.
+	metricClusterDeploymentProvisionUnderwaySeconds *prometheus.Desc
+}
+
+// collects the metrics for provisioningUnderwayCollector
+func (cc provisioningUnderwayCollector) Collect(ch chan<- prometheus.Metric) {
+	ccLog := log.WithField("controller", "metrics")
+	ccLog.Info("calculating provisioning underway metrics across all ClusterDeployments")
+
+	// Load all ClusterDeployments so we can accumulate facts about them.
+	clusterDeployments := &hivev1.ClusterDeploymentList{}
+	err := cc.client.List(context.Background(), clusterDeployments)
+	if err != nil {
+		log.WithError(err).Error("error listing cluster deployments")
+	} else {
+		for _, cd := range clusterDeployments.Items {
+			if cd.DeletionTimestamp == nil {
+				if !cd.Spec.Installed {
+					// Add install failure details for stuck provision
+					condition, reason := "Unknown", "Unknown"
+					for _, delayCondition := range provisioningDelayCondition {
+						if cdCondition := controllerutils.FindClusterDeploymentCondition(cd.Status.Conditions,
+							delayCondition); cdCondition != nil {
+							condition = string(delayCondition)
+							if cdCondition.Reason != "" {
+								reason = cdCondition.Reason
+							}
+							break
+						}
+					}
+
+					// For installing clusters we report the seconds since the cluster was created.
+					ch <- prometheus.MustNewConstMetric(
+						cc.metricClusterDeploymentProvisionUnderwaySeconds,
+						prometheus.GaugeValue,
```

Comment:
> https://prometheus.io/docs/concepts/metric_types/#gauge

`single numerical value that can arbitrarily go up and down.`

> https://prometheus.io/docs/concepts/metric_types/#counter

represents a single monotonically increasing counter whose value can only increase or be reset to zero on restart

shouldn't the time since start be a counter instead of gauge.

Reply:
I believe the increase referenced there is by one, the Inc() function, and there is not a way to increment to another value. I think gauge is correct here despite the fact we'll never go down.

Reply:
https://godoc.org/github.com/prometheus/client_golang/prometheus#Counter
has an Add(float64) interface

but i agree the guage has more ergonomic API for this usecase since it allows Set(float64)